source/module_elecstate/module_dm/density_matrix.cpp

```diff
@@ -484,8 +484,8 @@ void DensityMatrix<std::complex<double>, double>::cal_DMR()
             }
             for (int ir = 0; ir < tmp_ap.get_R_size(); ++ir)
             {
-                const int* r_index = tmp_ap.get_R_index(ir);
-                hamilt::BaseMatrix<double>* tmp_matrix = tmp_ap.find_matrix(r_index[0], r_index[1], r_index[2]);
+                const ModuleBase::Vector3<int> r_index = tmp_ap.get_R_index(ir);
```

Comment:
I suggest to refactor the function of `get_R_index(ir)` to 
```
const ModuleBase::Vector3<int>& r_index = tmp_ap.get_R_index(ir); // return value should be reference
```

Reply:
> I suggest to refactor the function of `get_R_index(ir)` to
> 
> ```
> const ModuleBase::Vector3<int>& r_index = tmp_ap.get_R_index(ir); // return value should be reference
> ```

This may not be a proper refactor. The function `get_R_index` also handles index errors and returns `ModuleBase::Vector3<int>(-1, -1, -1)` when the query index is out of bounds. Returning a reference is hard to manage this case.